{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data processing\n",
    "\n",
    "### 0 Introduction\n",
    "For this project we have 3 datasets:\n",
    "- **Proteomics:** Signal peptide abundances measured by analyzing the supernatant of fungal samples cultivated under two different media ccompositions (minimal media and minimal media with additional nitrogen)\n",
    "- **RNA sequence:** Gene expression levels obtained from transcriptome anlaysis of fungal samples cultivated under different conditions\n",
    "- **SignalP:** Potential signal peptide seuqences determined by using signalP algorithm to analyze the _A. oryzae_ genome \n",
    "\n",
    "The datasets have different structure and excess information that are not necessary for data analysis. This will make it difficult to connect the datasets during analysis. Hence, the datasets need to be processed to obtain a common structure which contains only the relevant information needed to carry out analysis and link between the datasets. \n",
    "\n",
    "### Agenda:\n",
    "As a rule of thumb, the datasets should have a generic structure which will satisfy:\n",
    "- Common name for the columns which contain same data content \n",
    "- Representative name that clearly identifies the data stored under the columns\n",
    "- Dataframe strucure that contains only the necessary information needed for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 Data\n",
    "\n",
    "### 1.1 Proteomics data\n",
    "Proteomics data consist of peptide sequence analysis of supernatants obtained from _Aspergillus_ samples that are grown under two different minimum media compositions (minimum media and minimum media with additional nitrogen). Each media composition was tested with 3 samples and the supernatants of each sample were analyzed through mass spectrometry, thus yielding the proteomics dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked</th>\n",
       "      <th>Protein FDR Confidence: Combined</th>\n",
       "      <th>Master</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Exp. q-value: Combined</th>\n",
       "      <th>Sum PEP Score</th>\n",
       "      <th>Coverage [%]</th>\n",
       "      <th># Peptides</th>\n",
       "      <th># PSMs</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 97</th>\n",
       "      <th>Unnamed: 98</th>\n",
       "      <th>Unnamed: 99</th>\n",
       "      <th>Unnamed: 100</th>\n",
       "      <th>Unnamed: 101</th>\n",
       "      <th>Unnamed: 102</th>\n",
       "      <th>Unnamed: 103</th>\n",
       "      <th>Unnamed: 104</th>\n",
       "      <th>Unnamed: 105</th>\n",
       "      <th>Unnamed: 106</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003000935-T-p1</td>\n",
       "      <td>transcript=AO090003000935-T | gene=AO090003000...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>599.687</td>\n",
       "      <td>89</td>\n",
       "      <td>31</td>\n",
       "      <td>7127</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Minimal medium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Minimal medium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090023000944-T-p1</td>\n",
       "      <td>transcript=AO090023000944-T | gene=AO090023000...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>523.784</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>3237</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sample 1</td>\n",
       "      <td>4.740753e+07</td>\n",
       "      <td>7.385813e+08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Raw</td>\n",
       "      <td>3.083450e+07</td>\n",
       "      <td>1.934537e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003001591-T-p1</td>\n",
       "      <td>transcript=AO090003001591-T | gene=AO090003001...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>519.742</td>\n",
       "      <td>68</td>\n",
       "      <td>33</td>\n",
       "      <td>2987</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sample 2</td>\n",
       "      <td>4.286160e+07</td>\n",
       "      <td>6.545026e+08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Scaled</td>\n",
       "      <td>4.676862e+01</td>\n",
       "      <td>5.097854e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>RFP_Fusion</td>\n",
       "      <td>RFP_Fusion</td>\n",
       "      <td>0.000</td>\n",
       "      <td>450.464</td>\n",
       "      <td>64</td>\n",
       "      <td>35</td>\n",
       "      <td>398</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sample 3</td>\n",
       "      <td>4.792621e+07</td>\n",
       "      <td>7.792201e+08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Normalized</td>\n",
       "      <td>4.606512e+07</td>\n",
       "      <td>2.786418e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090005001300-T-p1</td>\n",
       "      <td>transcript=AO090005001300-T | gene=AO090005001...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>384.472</td>\n",
       "      <td>90</td>\n",
       "      <td>45</td>\n",
       "      <td>554</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090001000075-T-p1</td>\n",
       "      <td>transcript=AO090001000075-T | gene=AO090001000...</td>\n",
       "      <td>0.044</td>\n",
       "      <td>1.795</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090005001355-T-p1</td>\n",
       "      <td>transcript=AO090005001355-T | gene=AO090005001...</td>\n",
       "      <td>0.046</td>\n",
       "      <td>1.790</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003000247-T-p1</td>\n",
       "      <td>transcript=AO090003000247-T | gene=AO090003000...</td>\n",
       "      <td>0.047</td>\n",
       "      <td>1.779</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090001000680-T-p1</td>\n",
       "      <td>transcript=AO090001000680-T | gene=AO090001000...</td>\n",
       "      <td>0.048</td>\n",
       "      <td>1.774</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090020000232-T-p1</td>\n",
       "      <td>transcript=AO090020000232-T | gene=AO090020000...</td>\n",
       "      <td>0.048</td>\n",
       "      <td>1.771</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>866 rows × 107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Checked Protein FDR Confidence: Combined          Master  \\\n",
       "0      False                             High  Master Protein   \n",
       "1      False                             High  Master Protein   \n",
       "2      False                             High  Master Protein   \n",
       "3      False                             High  Master Protein   \n",
       "4      False                             High  Master Protein   \n",
       "..       ...                              ...             ...   \n",
       "861    False                           Medium  Master Protein   \n",
       "862    False                           Medium  Master Protein   \n",
       "863    False                           Medium  Master Protein   \n",
       "864    False                           Medium  Master Protein   \n",
       "865    False                           Medium  Master Protein   \n",
       "\n",
       "               Accession                                        Description  \\\n",
       "0    AO090003000935-T-p1  transcript=AO090003000935-T | gene=AO090003000...   \n",
       "1    AO090023000944-T-p1  transcript=AO090023000944-T | gene=AO090023000...   \n",
       "2    AO090003001591-T-p1  transcript=AO090003001591-T | gene=AO090003001...   \n",
       "3             RFP_Fusion                                         RFP_Fusion   \n",
       "4    AO090005001300-T-p1  transcript=AO090005001300-T | gene=AO090005001...   \n",
       "..                   ...                                                ...   \n",
       "861  AO090001000075-T-p1  transcript=AO090001000075-T | gene=AO090001000...   \n",
       "862  AO090005001355-T-p1  transcript=AO090005001355-T | gene=AO090005001...   \n",
       "863  AO090003000247-T-p1  transcript=AO090003000247-T | gene=AO090003000...   \n",
       "864  AO090001000680-T-p1  transcript=AO090001000680-T | gene=AO090001000...   \n",
       "865  AO090020000232-T-p1  transcript=AO090020000232-T | gene=AO090020000...   \n",
       "\n",
       "     Exp. q-value: Combined  Sum PEP Score  Coverage [%]  # Peptides  # PSMs  \\\n",
       "0                     0.000        599.687            89          31    7127   \n",
       "1                     0.000        523.784            68          32    3237   \n",
       "2                     0.000        519.742            68          33    2987   \n",
       "3                     0.000        450.464            64          35     398   \n",
       "4                     0.000        384.472            90          45     554   \n",
       "..                      ...            ...           ...         ...     ...   \n",
       "861                   0.044          1.795             4           1       1   \n",
       "862                   0.046          1.790            14           1       1   \n",
       "863                   0.047          1.779             4           1       1   \n",
       "864                   0.048          1.774             1           1       1   \n",
       "865                   0.048          1.771             1           1       2   \n",
       "\n",
       "     ...  Unnamed: 97     Unnamed: 98  Unnamed: 99  Unnamed: 100  \\\n",
       "0    ...          NaN  Minimal medium          NaN           NaN   \n",
       "1    ...          NaN             NaN     Sample 1  4.740753e+07   \n",
       "2    ...          NaN             NaN     Sample 2  4.286160e+07   \n",
       "3    ...          NaN             NaN     Sample 3  4.792621e+07   \n",
       "4    ...          NaN             NaN          NaN           NaN   \n",
       "..   ...          ...             ...          ...           ...   \n",
       "861  ...          NaN             NaN          NaN           NaN   \n",
       "862  ...          NaN             NaN          NaN           NaN   \n",
       "863  ...          NaN             NaN          NaN           NaN   \n",
       "864  ...          NaN             NaN          NaN           NaN   \n",
       "865  ...          NaN             NaN          NaN           NaN   \n",
       "\n",
       "     Unnamed: 101  Unnamed: 102    Unnamed: 103 Unnamed: 104  Unnamed: 105  \\\n",
       "0             NaN           NaN  Minimal medium          NaN           NaN   \n",
       "1    7.385813e+08           NaN             NaN          Raw  3.083450e+07   \n",
       "2    6.545026e+08           NaN             NaN       Scaled  4.676862e+01   \n",
       "3    7.792201e+08           NaN             NaN   Normalized  4.606512e+07   \n",
       "4             NaN           NaN             NaN          NaN           NaN   \n",
       "..            ...           ...             ...          ...           ...   \n",
       "861           NaN           NaN             NaN          NaN           NaN   \n",
       "862           NaN           NaN             NaN          NaN           NaN   \n",
       "863           NaN           NaN             NaN          NaN           NaN   \n",
       "864           NaN           NaN             NaN          NaN           NaN   \n",
       "865           NaN           NaN             NaN          NaN           NaN   \n",
       "\n",
       "     Unnamed: 106  \n",
       "0             NaN  \n",
       "1    1.934537e+06  \n",
       "2    5.097854e+00  \n",
       "3    2.786418e+06  \n",
       "4             NaN  \n",
       "..            ...  \n",
       "861           NaN  \n",
       "862           NaN  \n",
       "863           NaN  \n",
       "864           NaN  \n",
       "865           NaN  \n",
       "\n",
       "[866 rows x 107 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the Excel file into a pandas dataframe \n",
    "df_proteomics = pd.read_excel('../Data/03_Proteomics_data/20221124_FJ_E1200_MWN_15cm_140min_500ng_#1572_proteins.xlsx')\n",
    "df_proteomics"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 RNA sequence data\n",
    "\n",
    "The RNA sequence data obtained from the survey of _Aspergillus oryzae_ RIB40 trancriptome analysis done by Wang B et al., (doi: 10.1093/nar/gkq256). This informaiton can be used to anlayse expression levels of the gene and normalize the peptide abundances observed in the proteomics data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Expression level (RPKM)</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Fraction of Coverage</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "      <th>Unnamed: 12</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "      <td>Liquid culture</td>\n",
       "      <td>Liquid culture under ER stress</td>\n",
       "      <td>Solid state culture</td>\n",
       "      <td>Solid state culture under ER stress</td>\n",
       "      <td>Liquid culture</td>\n",
       "      <td>Liquid culture under ER stress</td>\n",
       "      <td>Solid state culture</td>\n",
       "      <td>Solid state culture under ER stress</td>\n",
       "      <td>Gene description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090120000112</td>\n",
       "      <td>SC113</td>\n",
       "      <td>266998.0</td>\n",
       "      <td>267688.0</td>\n",
       "      <td>8388.947429</td>\n",
       "      <td>4557.36429</td>\n",
       "      <td>18205.843951</td>\n",
       "      <td>1067.257718</td>\n",
       "      <td>9966.433178</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peroxiredoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090120000080</td>\n",
       "      <td>SC113</td>\n",
       "      <td>171707.0</td>\n",
       "      <td>173307.0</td>\n",
       "      <td>4385.690027</td>\n",
       "      <td>4505.039311</td>\n",
       "      <td>1408.485967</td>\n",
       "      <td>7045.804664</td>\n",
       "      <td>4385.045891</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>translation elongation factor EF-1 alpha/Tu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090003001322</td>\n",
       "      <td>SC003</td>\n",
       "      <td>3582219.0</td>\n",
       "      <td>3583617.0</td>\n",
       "      <td>3790.864724</td>\n",
       "      <td>5548.417546</td>\n",
       "      <td>2682.238545</td>\n",
       "      <td>4221.991113</td>\n",
       "      <td>2836.683437</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>glyceraldehyde 3-phosphate dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090009000405</td>\n",
       "      <td>SC009</td>\n",
       "      <td>1073625.0</td>\n",
       "      <td>1074930.0</td>\n",
       "      <td>3154.906117</td>\n",
       "      <td>2223.300429</td>\n",
       "      <td>3199.479197</td>\n",
       "      <td>3920.072085</td>\n",
       "      <td>3152.671701</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mitochondrial ADP/ATP carrier proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11273</th>\n",
       "      <td>AO090005000065</td>\n",
       "      <td>SC005</td>\n",
       "      <td>168094.0</td>\n",
       "      <td>169221.0</td>\n",
       "      <td>0.042519</td>\n",
       "      <td>0.073975</td>\n",
       "      <td>0</td>\n",
       "      <td>0.063872</td>\n",
       "      <td>0.033124</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.04</td>\n",
       "      <td>predicted protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11274</th>\n",
       "      <td>AO090038000549</td>\n",
       "      <td>SC038</td>\n",
       "      <td>1488959.0</td>\n",
       "      <td>1489768.0</td>\n",
       "      <td>0.042205</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.164395</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>7-keto-8-aminopelargonate synthetase and relat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11275</th>\n",
       "      <td>AO090026000121</td>\n",
       "      <td>SC026</td>\n",
       "      <td>345917.0</td>\n",
       "      <td>348018.0</td>\n",
       "      <td>0.04217</td>\n",
       "      <td>0.040759</td>\n",
       "      <td>0</td>\n",
       "      <td>0.052789</td>\n",
       "      <td>0.073004</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>2-polyprenyl-6-methoxyphenol hydroxylase and r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11276</th>\n",
       "      <td>AO090001000066</td>\n",
       "      <td>SC001</td>\n",
       "      <td>146665.0</td>\n",
       "      <td>148342.0</td>\n",
       "      <td>0.042127</td>\n",
       "      <td>0</td>\n",
       "      <td>0.021312</td>\n",
       "      <td>0.098877</td>\n",
       "      <td>0.041023</td>\n",
       "      <td>0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.05</td>\n",
       "      <td>predicted protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11277</th>\n",
       "      <td>AO090010000539</td>\n",
       "      <td>SC010</td>\n",
       "      <td>1440458.0</td>\n",
       "      <td>1441531.0</td>\n",
       "      <td>0.042101</td>\n",
       "      <td>0.061039</td>\n",
       "      <td>0.113595</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11278 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene Contig      Start        End Expression level (RPKM)  \\\n",
       "0                 NaN    NaN        NaN        NaN                   Total   \n",
       "1      AO090120000112  SC113   266998.0   267688.0             8388.947429   \n",
       "2      AO090120000080  SC113   171707.0   173307.0             4385.690027   \n",
       "3      AO090003001322  SC003  3582219.0  3583617.0             3790.864724   \n",
       "4      AO090009000405  SC009  1073625.0  1074930.0             3154.906117   \n",
       "...               ...    ...        ...        ...                     ...   \n",
       "11273  AO090005000065  SC005   168094.0   169221.0                0.042519   \n",
       "11274  AO090038000549  SC038  1488959.0  1489768.0                0.042205   \n",
       "11275  AO090026000121  SC026   345917.0   348018.0                 0.04217   \n",
       "11276  AO090001000066  SC001   146665.0   148342.0                0.042127   \n",
       "11277  AO090010000539  SC010  1440458.0  1441531.0                0.042101   \n",
       "\n",
       "           Unnamed: 5                      Unnamed: 6           Unnamed: 7  \\\n",
       "0      Liquid culture  Liquid culture under ER stress  Solid state culture   \n",
       "1          4557.36429                    18205.843951          1067.257718   \n",
       "2         4505.039311                     1408.485967          7045.804664   \n",
       "3         5548.417546                     2682.238545          4221.991113   \n",
       "4         2223.300429                     3199.479197          3920.072085   \n",
       "...               ...                             ...                  ...   \n",
       "11273        0.073975                               0             0.063872   \n",
       "11274               0                               0                    0   \n",
       "11275        0.040759                               0             0.052789   \n",
       "11276               0                        0.021312             0.098877   \n",
       "11277        0.061039                        0.113595                    0   \n",
       "\n",
       "                                Unnamed: 8 Fraction of Coverage  \\\n",
       "0      Solid state culture under ER stress       Liquid culture   \n",
       "1                              9966.433178                    1   \n",
       "2                              4385.045891                    1   \n",
       "3                              2836.683437                    1   \n",
       "4                              3152.671701                    1   \n",
       "...                                    ...                  ...   \n",
       "11273                             0.033124                 0.09   \n",
       "11274                             0.164395                    0   \n",
       "11275                             0.073004                 0.05   \n",
       "11276                             0.041023                    0   \n",
       "11277                                    0                 0.07   \n",
       "\n",
       "                          Unnamed: 10          Unnamed: 11  \\\n",
       "0      Liquid culture under ER stress  Solid state culture   \n",
       "1                                   1                    1   \n",
       "2                                   1                    1   \n",
       "3                                   1                    1   \n",
       "4                                   1                    1   \n",
       "...                               ...                  ...   \n",
       "11273                               0                 0.09   \n",
       "11274                               0                    0   \n",
       "11275                               0                 0.07   \n",
       "11276                            0.03                 0.16   \n",
       "11277                            0.07                    0   \n",
       "\n",
       "                               Unnamed: 12  \\\n",
       "0      Solid state culture under ER stress   \n",
       "1                                        1   \n",
       "2                                        1   \n",
       "3                                        1   \n",
       "4                                        1   \n",
       "...                                    ...   \n",
       "11273                                 0.04   \n",
       "11274                                 0.22   \n",
       "11275                                 0.08   \n",
       "11276                                 0.05   \n",
       "11277                                    0   \n",
       "\n",
       "                                             Unnamed: 13  \n",
       "0                                       Gene description  \n",
       "1                                          peroxiredoxin  \n",
       "2            translation elongation factor EF-1 alpha/Tu  \n",
       "3               glyceraldehyde 3-phosphate dehydrogenase  \n",
       "4                 mitochondrial ADP/ATP carrier proteins  \n",
       "...                                                  ...  \n",
       "11273                                  predicted protein  \n",
       "11274  7-keto-8-aminopelargonate synthetase and relat...  \n",
       "11275  2-polyprenyl-6-methoxyphenol hydroxylase and r...  \n",
       "11276                                  predicted protein  \n",
       "11277                                  predicted protein  \n",
       "\n",
       "[11278 rows x 14 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the Excel file into a pandas dataframe\n",
    "df_rnaseq_oryzae = pd.read_excel('../Data/03_Proteomics_data/nar-02507-w-2009-File012.xlsx')\n",
    "df_rnaseq_oryzae"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 SignalP data\n",
    "\n",
    "SignalP is an algorithm developed in DTU which predicts the presence and location of signal peptide cleavage sites the amino acid sequences through analyzing the annotated protein profile of a specific organism. Hence, SignalP can be used to generate a list of potential signal peptide sequences. I would like to credit and emphasize my gratitude towards my supervisor Lucas Levassor who put the effor to generate the SignalP predictions, which were described in the notebook 00."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>signal_peptide_likelyhood</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090005000016-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.999803</td>\n",
       "      <td>MAPSHSFMLFLSVICTHLCSLVV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090005000029-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.999835</td>\n",
       "      <td>MHLRNIVIALAATAVASPVDLQD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005000042-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.999843</td>\n",
       "      <td>MKASFISRLLSLTAFAISSNLSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090005000053-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>0.854809</td>\n",
       "      <td>MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090005000059-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.999821</td>\n",
       "      <td>MHLQATLAVGLSLLGLTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>AO090103000483-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.833106</td>\n",
       "      <td>MKTSFLLAAIGFLYRLPCS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>AO090103000487-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.999710</td>\n",
       "      <td>MTRYLSFLFLLILFGNSVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>AO090103000493-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>MRGIVALSFLSVALGVT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>AO090701000994-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>MRLLLIAPLFSAVSYGAQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>AO090701001287-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>MLFYRAFGGLLCFLYACVTVS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene  start_pos  end_pos  signal_peptide_likelyhood  \\\n",
       "0     AO090005000016-T-p1          0       23                   0.999803   \n",
       "1     AO090005000029-T-p1          0       23                   0.999835   \n",
       "2     AO090005000042-T-p1          0       23                   0.999843   \n",
       "3     AO090005000053-T-p1          0       41                   0.854809   \n",
       "4     AO090005000059-T-p1          0       18                   0.999821   \n",
       "...                   ...        ...      ...                        ...   \n",
       "1056  AO090103000483-T-p1          0       19                   0.833106   \n",
       "1057  AO090103000487-T-p1          0       19                   0.999710   \n",
       "1058  AO090103000493-T-p1          0       17                   0.999791   \n",
       "1059  AO090701000994-T-p1          0       18                   0.999845   \n",
       "1060  AO090701001287-T-p1          0       21                   0.999866   \n",
       "\n",
       "                                       sequence  \n",
       "0                       MAPSHSFMLFLSVICTHLCSLVV  \n",
       "1                       MHLRNIVIALAATAVASPVDLQD  \n",
       "2                       MKASFISRLLSLTAFAISSNLSY  \n",
       "3     MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVAL  \n",
       "4                            MHLQATLAVGLSLLGLTL  \n",
       "...                                         ...  \n",
       "1056                        MKTSFLLAAIGFLYRLPCS  \n",
       "1057                        MTRYLSFLFLLILFGNSVF  \n",
       "1058                          MRGIVALSFLSVALGVT  \n",
       "1059                         MRLLLIAPLFSAVSYGAQ  \n",
       "1060                      MLFYRAFGGLLCFLYACVTVS  \n",
       "\n",
       "[1061 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the Excel file into a pandas dataframe\n",
    "df_signalP = pd.read_excel('../Data/01_SignalP_output/sigpep_predict.xlsx')\n",
    "df_signalP"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Data processing\n",
    "\n",
    "What are the parameters to isolate? What to keep and why to keep?\n",
    "\n",
    "### 2.1 Common changes in all datasets\n",
    "\n",
    "Among all the dataset, accession names serve as a common feature that can be used to connect the datasets, thus meaning all the datasets should be processed to have:\n",
    "- Common and representative name for the accession columns such as \"Accession\"\n",
    "- Common data structure which can be recognized such as accession numbers wihtout the suffix: \"-T-p1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the name of the accesion column in the dataframes into \"Accession\" \n",
    "df_signalP = df_signalP.rename(columns={'gene': 'Accession'})\n",
    "df_rnaseq_oryzae = df_rnaseq_oryzae.rename(columns={'Gene': 'Accession'})\n",
    "\n",
    "# Remove the suffix from the variables located in the \"Accession\" columns\n",
    "df_proteomics['Accession'] = df_proteomics['Accession'].str.replace('-T-p1', '')\n",
    "df_signalP['Accession'] = df_signalP['Accession'].str.replace('-T-p1', '')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Proteomics data processing\n",
    "\n",
    "In this dataset, some of the important processing steps that need to be taken include:\n",
    "\n",
    "- Removal of RFP_Fusion (Red Fluorescent Protein) as it is not part of the native protein pool of A. oryzae RIB40\n",
    "- Isolation of important parameters for analysis such as the columns:\n",
    "    - **Accession:** Unique identifier that is assigned to the signal peptide sequence \n",
    "    - **Abundace:** Raw signal peptide counts that are observed and recorded directly from mass spectrometry analysis \n",
    "    - **Abundances (Scaled):** Scaled version of raw singal peptide counts signal peptide counts that are observed from mass spectrometry analysis \n",
    "\n",
    "There are multiple columns with the name \"Abundance\" hence, distinctive patterns needs to be determined to isolate the specific columns. The data for minimum media (MM) and minimum media with additional nitrogen (MM_N) compositions are stored under the column names which show a specific naming pattern such as:\n",
    "\n",
    "- **Raw signal peptide abundance**\n",
    "    - Abundance: F16, Abundance: F17, Abundance: F18 - Minimum media (MM)\n",
    "    - Abundance: F28, Abundance: F29, Abundance: F30 - Minimum media with additional nitrogen (MM_N)\n",
    "- **Scaled signal peptide abundance**\n",
    "    - Abundances (Scaled): F16, Abundances (Scaled): F17, Abundances (Scaled): F18 - Minimum media (MM)\n",
    "    - Abundances (Scaled): F28, Abundances (Scaled): F29, Abundances (Scaled): F30 - Minimum media with additional nitrogen (MM_N)\n",
    "\n",
    "Hence, the \"Scaled\" and \"Raw\" abundace samples can be distinguished with the column pattern names:\n",
    "- **Pattern 1:** \"Abundance: and Scaled\"\n",
    "- **Pattern 2:** \"F16, F17, F18, F28, F29, F30\"\n",
    "\n",
    "Lets remove \"RFP_Fusion\" and create a boolean mask to select dataframe variables according to specific characters they contain in the dataframe such as using \"Scaled\" to select scaled abundances and \"Abundance:\" to select raw abundances "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Abundances (Scaled): F16: Sample, 1</th>\n",
       "      <th>Abundances (Scaled): F17: Sample, 2</th>\n",
       "      <th>Abundances (Scaled): F18: Sample, 3</th>\n",
       "      <th>Abundances (Scaled): F28: Sample, 13</th>\n",
       "      <th>Abundances (Scaled): F29: Sample, 14</th>\n",
       "      <th>Abundances (Scaled): F30: Sample, 15</th>\n",
       "      <th>Abundance: F16: Sample, 1</th>\n",
       "      <th>Abundance: F17: Sample, 2</th>\n",
       "      <th>Abundance: F18: Sample, 3</th>\n",
       "      <th>Abundance: F28: Sample, 13</th>\n",
       "      <th>Abundance: F29: Sample, 14</th>\n",
       "      <th>Abundance: F30: Sample, 15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090003000935</td>\n",
       "      <td>5.4</td>\n",
       "      <td>4.3</td>\n",
       "      <td>7.4</td>\n",
       "      <td>81.0</td>\n",
       "      <td>35.7</td>\n",
       "      <td>77.8</td>\n",
       "      <td>9.597858e+07</td>\n",
       "      <td>9.416572e+07</td>\n",
       "      <td>1.297344e+08</td>\n",
       "      <td>1.888527e+09</td>\n",
       "      <td>1.011602e+09</td>\n",
       "      <td>1.584488e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090023000944</td>\n",
       "      <td>242.7</td>\n",
       "      <td>250.1</td>\n",
       "      <td>258.5</td>\n",
       "      <td>42.9</td>\n",
       "      <td>72.6</td>\n",
       "      <td>45.7</td>\n",
       "      <td>2.378266e+08</td>\n",
       "      <td>3.016177e+08</td>\n",
       "      <td>2.504219e+08</td>\n",
       "      <td>5.514791e+07</td>\n",
       "      <td>1.134059e+08</td>\n",
       "      <td>5.128523e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090003001591</td>\n",
       "      <td>231.8</td>\n",
       "      <td>214.1</td>\n",
       "      <td>244.3</td>\n",
       "      <td>62.8</td>\n",
       "      <td>84.6</td>\n",
       "      <td>60.1</td>\n",
       "      <td>9.888853e+09</td>\n",
       "      <td>1.123887e+10</td>\n",
       "      <td>1.030518e+10</td>\n",
       "      <td>3.511919e+09</td>\n",
       "      <td>5.755309e+09</td>\n",
       "      <td>2.932830e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090005001300</td>\n",
       "      <td>96.3</td>\n",
       "      <td>38.4</td>\n",
       "      <td>17.0</td>\n",
       "      <td>89.7</td>\n",
       "      <td>111.3</td>\n",
       "      <td>82.3</td>\n",
       "      <td>7.204276e+07</td>\n",
       "      <td>3.532077e+07</td>\n",
       "      <td>1.257048e+07</td>\n",
       "      <td>8.793645e+07</td>\n",
       "      <td>1.328252e+08</td>\n",
       "      <td>7.046364e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090010000746</td>\n",
       "      <td>98.6</td>\n",
       "      <td>105.3</td>\n",
       "      <td>113.0</td>\n",
       "      <td>260.9</td>\n",
       "      <td>284.5</td>\n",
       "      <td>247.8</td>\n",
       "      <td>2.367148e+08</td>\n",
       "      <td>3.107989e+08</td>\n",
       "      <td>2.681958e+08</td>\n",
       "      <td>8.206836e+08</td>\n",
       "      <td>1.088923e+09</td>\n",
       "      <td>6.807444e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>AO090001000075</td>\n",
       "      <td>201.7</td>\n",
       "      <td>18.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>239.4</td>\n",
       "      <td>431.0</td>\n",
       "      <td>261.4</td>\n",
       "      <td>1.819305e+05</td>\n",
       "      <td>2.070978e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.830392e+05</td>\n",
       "      <td>6.199164e+05</td>\n",
       "      <td>2.698529e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>AO090005001355</td>\n",
       "      <td>64.6</td>\n",
       "      <td>146.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>184.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.104682e+04</td>\n",
       "      <td>5.864526e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.575187e+04</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>AO090003000247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>AO090001000680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.9</td>\n",
       "      <td>673.0</td>\n",
       "      <td>244.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.761538e+04</td>\n",
       "      <td>1.879930e+05</td>\n",
       "      <td>4.894831e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>AO090020000232</td>\n",
       "      <td>265.1</td>\n",
       "      <td>263.7</td>\n",
       "      <td>267.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.173873e+05</td>\n",
       "      <td>2.661302e+05</td>\n",
       "      <td>2.164992e+05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>865 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Accession  Abundances (Scaled): F16: Sample, 1  \\\n",
       "0    AO090003000935                                  5.4   \n",
       "1    AO090023000944                                242.7   \n",
       "2    AO090003001591                                231.8   \n",
       "4    AO090005001300                                 96.3   \n",
       "5    AO090010000746                                 98.6   \n",
       "..              ...                                  ...   \n",
       "861  AO090001000075                                201.7   \n",
       "862  AO090005001355                                 64.6   \n",
       "863  AO090003000247                                  NaN   \n",
       "864  AO090001000680                                  NaN   \n",
       "865  AO090020000232                                265.1   \n",
       "\n",
       "     Abundances (Scaled): F17: Sample, 2  Abundances (Scaled): F18: Sample, 3  \\\n",
       "0                                    4.3                                  7.4   \n",
       "1                                  250.1                                258.5   \n",
       "2                                  214.1                                244.3   \n",
       "4                                   38.4                                 17.0   \n",
       "5                                  105.3                                113.0   \n",
       "..                                   ...                                  ...   \n",
       "861                                 18.7                                  NaN   \n",
       "862                                146.2                                  NaN   \n",
       "863                                  NaN                                  NaN   \n",
       "864                                  NaN                                  NaN   \n",
       "865                                263.7                                267.0   \n",
       "\n",
       "     Abundances (Scaled): F28: Sample, 13  \\\n",
       "0                                    81.0   \n",
       "1                                    42.9   \n",
       "2                                    62.8   \n",
       "4                                    89.7   \n",
       "5                                   260.9   \n",
       "..                                    ...   \n",
       "861                                 239.4   \n",
       "862                                   NaN   \n",
       "863                                   NaN   \n",
       "864                                 250.9   \n",
       "865                                   NaN   \n",
       "\n",
       "     Abundances (Scaled): F29: Sample, 14  \\\n",
       "0                                    35.7   \n",
       "1                                    72.6   \n",
       "2                                    84.6   \n",
       "4                                   111.3   \n",
       "5                                   284.5   \n",
       "..                                    ...   \n",
       "861                                 431.0   \n",
       "862                                 184.2   \n",
       "863                                   NaN   \n",
       "864                                 673.0   \n",
       "865                                   NaN   \n",
       "\n",
       "     Abundances (Scaled): F30: Sample, 15  Abundance: F16: Sample, 1  \\\n",
       "0                                    77.8               9.597858e+07   \n",
       "1                                    45.7               2.378266e+08   \n",
       "2                                    60.1               9.888853e+09   \n",
       "4                                    82.3               7.204276e+07   \n",
       "5                                   247.8               2.367148e+08   \n",
       "..                                    ...                        ...   \n",
       "861                                 261.4               1.819305e+05   \n",
       "862                                   NaN               2.104682e+04   \n",
       "863                                   NaN                        NaN   \n",
       "864                                 244.1                        NaN   \n",
       "865                                   NaN               2.173873e+05   \n",
       "\n",
       "     Abundance: F17: Sample, 2  Abundance: F18: Sample, 3  \\\n",
       "0                 9.416572e+07               1.297344e+08   \n",
       "1                 3.016177e+08               2.504219e+08   \n",
       "2                 1.123887e+10               1.030518e+10   \n",
       "4                 3.532077e+07               1.257048e+07   \n",
       "5                 3.107989e+08               2.681958e+08   \n",
       "..                         ...                        ...   \n",
       "861               2.070978e+04                        NaN   \n",
       "862               5.864526e+04                        NaN   \n",
       "863                        NaN                        NaN   \n",
       "864                        NaN                        NaN   \n",
       "865               2.661302e+05               2.164992e+05   \n",
       "\n",
       "     Abundance: F28: Sample, 13  Abundance: F29: Sample, 14  \\\n",
       "0                  1.888527e+09                1.011602e+09   \n",
       "1                  5.514791e+07                1.134059e+08   \n",
       "2                  3.511919e+09                5.755309e+09   \n",
       "4                  8.793645e+07                1.328252e+08   \n",
       "5                  8.206836e+08                1.088923e+09   \n",
       "..                          ...                         ...   \n",
       "861                2.830392e+05                6.199164e+05   \n",
       "862                         NaN                9.575187e+04   \n",
       "863                         NaN                         NaN   \n",
       "864                5.761538e+04                1.879930e+05   \n",
       "865                         NaN                         NaN   \n",
       "\n",
       "     Abundance: F30: Sample, 15  \n",
       "0                  1.584488e+09  \n",
       "1                  5.128523e+07  \n",
       "2                  2.932830e+09  \n",
       "4                  7.046364e+07  \n",
       "5                  6.807444e+08  \n",
       "..                          ...  \n",
       "861                2.698529e+05  \n",
       "862                         NaN  \n",
       "863                         NaN  \n",
       "864                4.894831e+04  \n",
       "865                         NaN  \n",
       "\n",
       "[865 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove RFP_Fusion from the dataframe as it is not part of the native protein pool\n",
    "df_proteomics = df_proteomics[df_proteomics.Accession != 'RFP_Fusion']\n",
    "\n",
    "# Create a boolean mask that is True for columns that contain the pattaern1: \"Scaled, Abundance:\" and pattern2: \"F16, F17, F18, F28, F29; F30\"\n",
    "pattern1 = 'Abundance:|Scaled'\n",
    "pattern2 = 'F16|F17|F18|F28|F29|F30'\n",
    "df_raw_scaled_abundances = df_proteomics.loc[:, df_proteomics.columns.str.contains(pattern1) & df_proteomics.columns.str.contains(pattern2, regex=True)]\n",
    "\n",
    "# Recall the dataframe to combine accession with ioslated list of columns for the peptide abundances \n",
    "df_proteomics_processed = df_proteomics[['Accession']+ list(df_raw_scaled_abundances.columns)]\n",
    "df_proteomics_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and save a summary dataframe\n",
    "df_proteomics_processed.to_csv('../Data/04_Processed_data/proteomics_processed.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 RNA sequence data processing\n",
    "\n",
    "This dataset had problem with the naming of the columns due to the excel document structure which caused some of the coulumn names to be moved to the first row leading \"Unnamed:\" annotation of the columns.\n",
    "\n",
    "In this dataset, some of the important processing steps that need to be taken include:\n",
    "- Change the column names \"Unnamed: 5\" to a representative name and cut the first row\n",
    "- Isolation of important parameters for analysis such as the columns:\n",
    "    - **Accession:** Unique identifier that is assigned to the signal peptide sequence \n",
    "    - **Liquid culture:** Gene expression levels observed under liquid culture cultivation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Liquid culture rna expression levels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090120000112</td>\n",
       "      <td>4557.36429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090120000080</td>\n",
       "      <td>4505.039311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090003001322</td>\n",
       "      <td>5548.417546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090009000405</td>\n",
       "      <td>2223.300429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000661</td>\n",
       "      <td>2355.882643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11273</th>\n",
       "      <td>AO090005000065</td>\n",
       "      <td>0.073975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11274</th>\n",
       "      <td>AO090038000549</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11275</th>\n",
       "      <td>AO090026000121</td>\n",
       "      <td>0.040759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11276</th>\n",
       "      <td>AO090001000066</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11277</th>\n",
       "      <td>AO090010000539</td>\n",
       "      <td>0.061039</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11277 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Accession Liquid culture rna expression levels\n",
       "1      AO090120000112                           4557.36429\n",
       "2      AO090120000080                          4505.039311\n",
       "3      AO090003001322                          5548.417546\n",
       "4      AO090009000405                          2223.300429\n",
       "5      AO090003000661                          2355.882643\n",
       "...               ...                                  ...\n",
       "11273  AO090005000065                             0.073975\n",
       "11274  AO090038000549                                    0\n",
       "11275  AO090026000121                             0.040759\n",
       "11276  AO090001000066                                    0\n",
       "11277  AO090010000539                             0.061039\n",
       "\n",
       "[11277 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename the column \"Unnamed: 5\" into a representative name\n",
    "df_rnaseq_oryzae = df_rnaseq_oryzae.rename(columns={'Unnamed: 5': 'Liquid culture rna expression levels'})\n",
    "\n",
    "# Isolate the important parameters \n",
    "df_rnaseq_oryzae_processed = df_rnaseq_oryzae[['Accession', 'Liquid culture rna expression levels']]\n",
    "\n",
    "# Drop the first row as it has been used for column names\n",
    "df_rnaseq_oryzae_processed = df_rnaseq_oryzae_processed.drop(df_rnaseq_oryzae.index[0])\n",
    "df_rnaseq_oryzae_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and save a summary dataframe\n",
    "df_rnaseq_oryzae_processed.to_csv('../Data/04_Processed_data/rnaseq_oryzae_processed.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 SignalP data processing\n",
    "\n",
    "In this dataset, some of the important processing steps that need to be taken include:\n",
    "- Change the column names \"end_pos\" to a representative name\n",
    "- Isolation of important parameters for analysis such as the columns:\n",
    "    - **Accession:** Unique identifier that is assigned to the signal peptide sequence \n",
    "    - **end_pos:** Length of the predicted signal peptides\n",
    "    - **sequence:** Amino acid sequence of the predicfted signal peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090005000016</td>\n",
       "      <td>23</td>\n",
       "      <td>MAPSHSFMLFLSVICTHLCSLVV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090005000029</td>\n",
       "      <td>23</td>\n",
       "      <td>MHLRNIVIALAATAVASPVDLQD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005000042</td>\n",
       "      <td>23</td>\n",
       "      <td>MKASFISRLLSLTAFAISSNLSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090005000053</td>\n",
       "      <td>41</td>\n",
       "      <td>MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090005000059</td>\n",
       "      <td>18</td>\n",
       "      <td>MHLQATLAVGLSLLGLTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>AO090103000483</td>\n",
       "      <td>19</td>\n",
       "      <td>MKTSFLLAAIGFLYRLPCS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>AO090103000487</td>\n",
       "      <td>19</td>\n",
       "      <td>MTRYLSFLFLLILFGNSVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>AO090103000493</td>\n",
       "      <td>17</td>\n",
       "      <td>MRGIVALSFLSVALGVT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>AO090701000994</td>\n",
       "      <td>18</td>\n",
       "      <td>MRLLLIAPLFSAVSYGAQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>AO090701001287</td>\n",
       "      <td>21</td>\n",
       "      <td>MLFYRAFGGLLCFLYACVTVS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Accession  length                                   sequence\n",
       "0     AO090005000016      23                    MAPSHSFMLFLSVICTHLCSLVV\n",
       "1     AO090005000029      23                    MHLRNIVIALAATAVASPVDLQD\n",
       "2     AO090005000042      23                    MKASFISRLLSLTAFAISSNLSY\n",
       "3     AO090005000053      41  MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVAL\n",
       "4     AO090005000059      18                         MHLQATLAVGLSLLGLTL\n",
       "...              ...     ...                                        ...\n",
       "1056  AO090103000483      19                        MKTSFLLAAIGFLYRLPCS\n",
       "1057  AO090103000487      19                        MTRYLSFLFLLILFGNSVF\n",
       "1058  AO090103000493      17                          MRGIVALSFLSVALGVT\n",
       "1059  AO090701000994      18                         MRLLLIAPLFSAVSYGAQ\n",
       "1060  AO090701001287      21                      MLFYRAFGGLLCFLYACVTVS\n",
       "\n",
       "[1061 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename the column \"Unnamed: 5\" into a representative name\n",
    "df_signalP = df_signalP.rename(columns={'end_pos': 'length'})\n",
    "\n",
    "# Isolate the important parameters\n",
    "df_signalP_processed = df_signalP[['Accession', 'length', 'sequence']]\n",
    "df_signalP_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and save a summary dataframe\n",
    "df_signalP_processed.to_csv('../Data/04_Processed_data/signalP_processed.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv_spep",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
