{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["%%capture\n","!pip install pandas\n","!pip install matplotlib\n","!pip install seaborn\n","!pip install umap-learn\n","!pip install hdbscan\n","!pip install biopython\n"]},{"cell_type":"code","execution_count":2,"metadata":{"id":"oNpjvcwG1Kql"},"outputs":[],"source":["import pandas as pd"]},{"cell_type":"code","execution_count":3,"metadata":{"id":"vEuJFloX2C9t"},"outputs":[],"source":["df = pd.read_csv('../data/02_all_signal_peptides/biochemical_result_w_abundance.csv', index_col=False)"]},{"cell_type":"code","execution_count":4,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1705399148648,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"1D6xWLTj2IAE","outputId":"3782a41c-8888-4540-f52b-3449c1c35a3a"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>gene</th>\n","      <th>n_relative_length</th>\n","      <th>n_kd_hydrophobicity</th>\n","      <th>n_charge_at_ph7</th>\n","      <th>h_relative_length</th>\n","      <th>h_kd_hydrophobicity</th>\n","      <th>h_charge_at_ph7</th>\n","      <th>c_relative_length</th>\n","      <th>c_kd_hydrophobicity</th>\n","      <th>c_charge_at_ph7</th>\n","      <th>abundance</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>AO090005000016-T-p1</td>\n","      <td>4</td>\n","      <td>0.325000</td>\n","      <td>-0.799151</td>\n","      <td>16</td>\n","      <td>1.582353</td>\n","      <td>-0.586943</td>\n","      <td>2</td>\n","      <td>3.400000</td>\n","      <td>-0.590986</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>AO090005000029-T-p1</td>\n","      <td>2</td>\n","      <td>-0.650000</td>\n","      <td>-0.775725</td>\n","      <td>10</td>\n","      <td>1.590909</td>\n","      <td>0.442738</td>\n","      <td>10</td>\n","      <td>-0.145455</td>\n","      <td>-1.505143</td>\n","      <td>0.000033</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>AO090005000042-T-p1</td>\n","      <td>2</td>\n","      <td>-1.000000</td>\n","      <td>0.196884</td>\n","      <td>18</td>\n","      <td>1.184211</td>\n","      <td>0.494293</td>\n","      <td>2</td>\n","      <td>-0.833333</td>\n","      <td>-0.827738</td>\n","      <td>0.000048</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>AO090005000053-T-p1</td>\n","      <td>26</td>\n","      <td>0.684615</td>\n","      <td>2.192839</td>\n","      <td>9</td>\n","      <td>1.660000</td>\n","      <td>-0.572083</td>\n","      <td>5</td>\n","      <td>1.116667</td>\n","      <td>-0.147963</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>AO090005000059-T-p1</td>\n","      <td>2</td>\n","      <td>-0.650000</td>\n","      <td>-0.775725</td>\n","      <td>13</td>\n","      <td>1.771429</td>\n","      <td>-0.557223</td>\n","      <td>2</td>\n","      <td>1.633333</td>\n","      <td>-0.857574</td>\n","      <td>0.000000</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                  gene  n_relative_length  n_kd_hydrophobicity  \\\n","0  AO090005000016-T-p1                  4             0.325000   \n","1  AO090005000029-T-p1                  2            -0.650000   \n","2  AO090005000042-T-p1                  2            -1.000000   \n","3  AO090005000053-T-p1                 26             0.684615   \n","4  AO090005000059-T-p1                  2            -0.650000   \n","\n","   n_charge_at_ph7  h_relative_length  h_kd_hydrophobicity  h_charge_at_ph7  \\\n","0        -0.799151                 16             1.582353        -0.586943   \n","1        -0.775725                 10             1.590909         0.442738   \n","2         0.196884                 18             1.184211         0.494293   \n","3         2.192839                  9             1.660000        -0.572083   \n","4        -0.775725                 13             1.771429        -0.557223   \n","\n","   c_relative_length  c_kd_hydrophobicity  c_charge_at_ph7  abundance  \n","0                  2             3.400000        -0.590986   0.000000  \n","1                 10            -0.145455        -1.505143   0.000033  \n","2                  2            -0.833333        -0.827738   0.000048  \n","3                  5             1.116667        -0.147963   0.000000  \n","4                  2             1.633333        -0.857574   0.000000  "]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["df.head()\n"]},{"cell_type":"markdown","metadata":{"id":"8Z4xUCKR5_wP"},"source":["# UMAP calculation"]},{"cell_type":"code","execution_count":5,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":36644,"status":"ok","timestamp":1705399277029,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"ktawdilQ5-NS","outputId":"826d86bd-02eb-4156-8900-f9d0ac5dd88e"},"outputs":[],"source":["import pandas as pd\n","import umap.umap_ as umap\n","import matplotlib.pyplot as plt\n","from sklearn.preprocessing import StandardScaler\n","import hdbscan\n","import seaborn as sns\n","import matplotlib.colors as mcolors\n","\n","\n","# Prepare the DataFrame by dropping non-numeric columns\n","data_for_umap = df.drop(columns=['gene', 'abundance'])\n","\n","# Standardize the features\n","scaler = StandardScaler()\n","scaled_data = scaler.fit_transform(data_for_umap)\n","\n","# Apply UMAP\n","umap_model = umap.UMAP(n_neighbors=15, min_dist=0.1, n_components=2, random_state=1234)\n","umap_results = umap_model.fit_transform(scaled_data)"]},{"cell_type":"markdown","metadata":{"id":"3UpVZque5Jzn"},"source":["## UMAP clustering with hdbscan"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":733},"executionInfo":{"elapsed":2103,"status":"ok","timestamp":1705399279130,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"lHJ7EjGH2Iwt","outputId":"629d06fb-d435-4cae-d7c1-a3d64ced3d2f"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Set the aesthetic style of the plots\n","sns.set(style='white', context='notebook')\n","\n","\n","# Apply HDBSCAN clustering\n","clusterer = hdbscan.HDBSCAN(min_cluster_size=50, gen_min_span_tree=True)\n","cluster_labels = clusterer.fit_predict(umap_results)\n","\n","# Generating a custom discrete colormap\n","cmap = plt.cm.Spectral\n","norm = mcolors.Normalize(vmin=cluster_labels.min(), vmax=cluster_labels.max())\n","\n","# Visualize the results with clusters\n","plt.figure(figsize=(10, 8))\n","scatter = plt.scatter(umap_results[:, 0], umap_results[:, 1], c=cluster_labels, cmap=cmap, norm=norm, s=50, alpha=0.8, edgecolor='k')\n","\n","# Adding a color bar\n","plt.colorbar(scatter, label='Cluster', shrink=0.8)\n","\n","# Enhancing plot aesthetics\n","plt.title('UMAP Projection with HDBSCAN Clustering', fontsize=16, fontweight='bold')\n","plt.xlabel('UMAP Component 1', fontsize=14)\n","plt.ylabel('UMAP Component 2', fontsize=14)\n","plt.xticks(fontsize=12)\n","plt.yticks(fontsize=12)\n","plt.savefig('../data/plots/UMAP_hdbscan_clustering.png', dpi=300, bbox_inches='tight')\n","\n","\n","plt.show()\n"]},{"cell_type":"markdown","metadata":{"id":"Zgim-n455S9a"},"source":["## UMAP with K-means"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":768},"executionInfo":{"elapsed":1247,"status":"ok","timestamp":1705399280375,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"jWw8V8B_5IW3","outputId":"ff45864b-0eda-487a-d88a-606e992d5d2a"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["import matplotlib.pyplot as plt\n","import seaborn as sns\n","from sklearn.cluster import KMeans\n","import umap.umap_ as umap\n","from sklearn.preprocessing import StandardScaler\n","\n","NUMBER_OF_CLUSTERS = 3\n","\n","# Apply K-means clustering\n","kmeans = KMeans(n_clusters=NUMBER_OF_CLUSTERS, random_state=1234)  # Adjust the number of clusters as needed\n","kmeans_clusters = kmeans.fit_predict(umap_results)\n","\n","# Set the aesthetic style of the plots\n","sns.set(style='white', context='notebook')\n","\n","# Create a color palette\n","palette = sns.color_palette(\"bright\", NUMBER_OF_CLUSTERS)  # Adjust the number of colors based on clusters\n","\n","# Visualize the results with clusters\n","plt.figure(figsize=(10, 8))\n","scatter = plt.scatter(umap_results[:, 0], umap_results[:, 1], c=kmeans_clusters, cmap='viridis', s=50, alpha=0.8, edgecolor='k')\n","\n","# Adding a color bar\n","plt.colorbar(scatter, label='Cluster', shrink=0.8)\n","\n","# Enhancing plot aesthetics\n","plt.title('UMAP Projection with K-Means Clustering', fontsize=16, fontweight='bold')\n","plt.xlabel('UMAP Component 1', fontsize=14)\n","plt.ylabel('UMAP Component 2', fontsize=14)\n","plt.xticks(fontsize=12)\n","plt.yticks(fontsize=12)\n","plt.savefig('../data/plots/UMAP_kmeans_clustering.png', dpi=300, bbox_inches='tight')\n","\n","\n","plt.show()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1705399280375,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"9cQSAtR33Ch9","outputId":"1721439b-daa0-4ccf-b3a1-94ca3349f272"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Add the cluster assignments to your original DataFrame\n","df['umap_cluster'] = kmeans_clusters\n","\n","# Now you can see each row with its corresponding cluster\n","df.head()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"mZ0UbvOm6p8u"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["\n","# If you want to see rows belonging to a specific cluster, you can filter the DataFrame\n","cluster_number = 0  # Change this to the cluster number you're interested in\n","specific_cluster_df = df[df['umap_cluster'] == cluster_number]\n","specific_cluster_df.to_csv('../data/02_all_signal_peptides/groups_of_sps/group_data1.csv')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"SGLsguvA6uEk"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["cluster_number = 1  # Change this to the cluster number you're interested in\n","specific_cluster_df1 = df[df['umap_cluster'] == cluster_number]\n","specific_cluster_df1.to_csv('../data/02_all_signal_peptides/groups_of_sps/group_data2')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GM5BV7vt68yr"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["cluster_number = 2  # Change this to the cluster number you're interested in\n","specific_cluster_df2 = df[df['umap_cluster'] == cluster_number]\n","specific_cluster_df2.to_csv('../data/02_all_signal_peptides/groups_of_sps/group_data3')"]},{"cell_type":"markdown","metadata":{"id":"E5yd0ExA8sxI"},"source":["## PCA and t-SNE"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":669},"executionInfo":{"elapsed":9264,"status":"ok","timestamp":1705399289636,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"-9fHh4YU8vYb","outputId":"c28d925d-567b-4666-fe56-389c49323a12"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","from sklearn.preprocessing import StandardScaler\n","from sklearn.decomposition import PCA\n","from sklearn.manifold import TSNE\n","from sklearn.cluster import KMeans\n","\n","# Assuming 'df' is your DataFrame\n","\n","# Prepare the DataFrame by dropping the non-numeric column\n","data_for_analysis = df.drop(columns=['gene', 'abundance'])\n","\n","# Standardize the features\n","scaler = StandardScaler()\n","scaled_data = scaler.fit_transform(data_for_analysis)\n","\n","# Apply PCA\n","pca = PCA(n_components=2)\n","pca_results = pca.fit_transform(scaled_data)\n","\n","# Apply t-SNE\n","tsne = TSNE(n_components=2, perplexity=30, learning_rate=200, random_state=1234)\n","tsne_results = tsne.fit_transform(scaled_data)\n","\n","# Apply K-means clustering\n","n_clusters = 3  # Adjust based on your data\n","kmeans = KMeans(n_clusters=n_clusters, random_state=1234)\n","pca_clusters = kmeans.fit_predict(pca_results)\n","tsne_clusters = kmeans.fit_predict(tsne_results)\n","\n","# Set style for plots\n","sns.set(style='white', context='notebook')\n","\n","# Plot PCA results\n","plt.figure(figsize=(12, 6))\n","plt.subplot(1, 2, 1)\n","plt.scatter(pca_results[:, 0], pca_results[:, 1], c=pca_clusters, cmap='viridis', s=50, alpha=0.6, edgecolor='k')\n","plt.title('PCA with K-means Clustering', fontsize=14, fontweight='bold')\n","plt.xlabel('PCA Component 1', fontsize=12)\n","plt.ylabel('PCA Component 2', fontsize=12)\n","\n","# Plot t-SNE results\n","plt.subplot(1, 2, 2)\n","plt.scatter(tsne_results[:, 0], tsne_results[:, 1], c=tsne_clusters, cmap='viridis', s=50, alpha=0.6, edgecolor='k')\n","plt.title('t-SNE with K-means Clustering', fontsize=14, fontweight='bold')\n","plt.xlabel('t-SNE Component 1', fontsize=12)\n","plt.ylabel('t-SNE Component 2', fontsize=12)\n","\n","# Adjust layout\n","plt.tight_layout()\n","\n","# Save the figure (optional)\n","plt.savefig('../data/plots/PCA_tSNE_Clustering.png', dpi=300, bbox_inches='tight')\n","\n","plt.show()\n"]},{"cell_type":"markdown","metadata":{"id":"mXhy51vG-406"},"source":["## Comparing clusters"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"w4rOggRE8vsK"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Add the t-SNE clusters to your DataFrame\n","df['tsne_cluster'] = tsne_clusters"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":175},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1705399289636,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"_Xl8ioM2-v7q","outputId":"6a99db3f-bd39-44b3-c04a-54800a22fca7"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["crosstab = pd.crosstab(df['tsne_cluster'], df['umap_cluster'])\n","crosstab"]},{"cell_type":"markdown","metadata":{"id":"SVM-ASj_ABQL"},"source":["## Adjusted Rand Index (ARI) or Normalized Mutual Information (NMI)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1705399289637,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"6SRihNKB_FVm","outputId":"856ed3ba-7d10-41ca-b18d-0f8c9cd7ba53"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["from sklearn.metrics import adjusted_rand_score, normalized_mutual_info_score\n","\n","# Calculate Adjusted Rand Index\n","ari = adjusted_rand_score(df['tsne_cluster'], df['umap_cluster'])\n","\n","# Calculate Normalized Mutual Information\n","nmi = normalized_mutual_info_score(df['tsne_cluster'], df['umap_cluster'])\n","\n","print(f\"Adjusted Rand Index: {ari:.3f}\")\n","print(f\"Normalized Mutual Information: {nmi:.3f}\")\n"]},{"cell_type":"markdown","metadata":{"id":"HeugbmtFt9yj"},"source":["## Filtering out gene ids"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":556},"executionInfo":{"elapsed":7,"status":"ok","timestamp":1705399511612,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"rabIXNkYuBjm","outputId":"e19b26b7-3c39-42ec-8613-1e8bb4df8924"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["martis_list = [\n","    \"AO090003001591-T-p1\",\n","    \"AO090005001090-T-p1\",\n","    \"AO090023000905-T-p1\",\n","    \"AO090038000279-T-p1\",\n","    \"AO090701000295-T-p1\",\n","    \"AO090003001036-T-p1\",\n","    \"AO090001000635-T-p1\",\n","    \"AO090005000409-T-p1\",\n","    \"AO090009000117-T-p1\",\n","    \"AO090102000440-T-p1\"\n","]\n","\n","# Retaining the rows with these gene IDs\n","retained_df = df[df['gene'].isin(martis_list)]\n","retained_df"]},{"cell_type":"markdown","metadata":{"id":"eZY4rtt0ECKS"},"source":["\n","### Understanding ARI and NMI:\n","\n","- **Adjusted Rand Index (ARI)**: Measures the similarity of the two assignments, ignoring permutations and with chance normalization. An ARI score of 1 means perfect labeling alignment, while 0 or negative values indicate no better than random chance.\n","\n","- **Normalized Mutual Information (NMI)**: Also measures the similarity, but it's normalized to account for chance. Like ARI, a score of 1 indicates perfect agreement, and lower values indicate less agreement.\n"]},{"cell_type":"markdown","metadata":{"id":"Ec7LmAMTk1Ga"},"source":["# Looking at the 3 groupings of Signal peptides through DEEPloc"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Sample DataFrame loading (replace with actual DataFrame loading as needed)\n","df1_deeploc = pd.read_csv('../data/03_proteomics_data/deeploc_predictions/all_sps/results_646764CA00004786815A8068.csv')\n","df2_deeploc = pd.read_csv('../data/03_proteomics_data/deeploc_predictions/all_sps/results_6467637F0000328EBD10D7B6.csv')\n","df3_deeploc = pd.read_csv('../data/03_proteomics_data/deeploc_predictions/all_sps/results_6467641D00003C072DBA1C82.csv')\n","df_deeploc_data = pd.concat([df1_deeploc, df2_deeploc, df3_deeploc])\n","df_deeploc_data = df_deeploc_data.rename(columns ={'Protein_ID':'gene'})\n","df_deeploc_data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["cluster1 = specific_cluster_df.merge(df_deeploc_data,on='gene')\n","cluster2 = specific_cluster_df1.merge(df_deeploc_data,on='gene')\n","cluster3 = specific_cluster_df2.merge(df_deeploc_data,on='gene')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","\n","# Function to get localization counts\n","def get_localization_counts(df):\n","    return df['Localizations'].value_counts()\n","\n","# Get counts for each cluster\n","counts1 = get_localization_counts(cluster1)\n","counts2 = get_localization_counts(cluster2)\n","counts3 = get_localization_counts(cluster3)\n","\n","# Set the aesthetics for the plots\n","sns.set(style=\"whitegrid\")\n","plt.rcParams.update({'font.size': 10, 'font.family': 'sans-serif'})\n","\n","# Create subplots with larger figure size\n","fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(25, 8), sharey=True)\n","\n","# Define a color palette with enough colors\n","max_unique_values = max(counts1.size, counts2.size, counts3.size)\n","palette = sns.color_palette(\"husl\", max_unique_values)\n","\n","# Function to create a barplot for a cluster\n","def create_barplot(ax, counts, title):\n","    sns.barplot(ax=ax, x=counts.index, y=counts.values, palette=palette[:counts.size])\n","    ax.set_title(title, fontsize=14, fontweight='bold')\n","    ax.set_xlabel('Localizations', fontsize=12)\n","    ax.set_ylabel('Count', fontsize=12)\n","    ax.tick_params(axis='x', rotation=90)  # Rotate labels\n","\n","# Create bar plots for each cluster\n","create_barplot(axes[0], counts1, 'Cluster 1 Localizations')\n","create_barplot(axes[1], counts2, 'Cluster 2 Localizations')\n","create_barplot(axes[2], counts3, 'Cluster 3 Localizations')\n","\n","# Adjust layout and spacing\n","plt.tight_layout()\n","plt.subplots_adjust(wspace=0.1)  # Adjust horizontal spacing\n","plt.show()\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["cluster1"]},{"cell_type":"markdown","metadata":{"id":"v4OEb6VaF7-X"},"source":["## Gene ontology of different groups of SPs"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wnWuwMYAUvFy"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["cluster1 = list(specific_cluster_df['gene'])\n","cluster2 = list(specific_cluster_df1['gene'])\n","cluster3 = list(specific_cluster_df2['gene'])"]},{"cell_type":"markdown","metadata":{"id":"TB12nFW5VArO"},"source":["These cluster were sent to FUNGIDB and go annotation was made:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":391},"executionInfo":{"elapsed":451,"status":"error","timestamp":1705399290084,"user":{"displayName":"Lucas Levassor","userId":"00761723138693003107"},"user_tz":-60},"id":"iETsDkYEmCdo","outputId":"7c4b6195-9c35-4460-d892-ef1b340564e5"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Sample DataFrame loading (replace with actual DataFrame loading as needed)\n","df1 = pd.read_csv('../data/14_GO_analysis/hiddenGoEnrichmentResult_group1.tsv', sep='\\t')\n","df2 = pd.read_csv('../data/14_GO_analysis/hiddenGoEnrichmentResult_group2.tsv', sep='\\t')\n","df3 = pd.read_csv('../data/14_GO_analysis/hiddenGoEnrichmentResult_group3.tsv', sep='\\t')\n","df1"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TgFs-xiMU0cb"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","\n","\n","# Assuming df1, df2, df3 are your dataframes\n","\n","def plot_top_bgd_genes(df, title, top_n=10):\n","    # Sort the dataframe by 'Result count' and select top N\n","    df_sorted = df.sort_values(by='Result count', ascending=False).head(top_n)\n","\n","    # Setting the aesthetic style for the plots\n","    sns.set(style='whitegrid', context='talk', palette='viridis')\n","\n","    # Plotting\n","    plt.figure(figsize=(8, 4))\n","    sns.barplot(x='Result count', y='Name', data=df_sorted, edgecolor='black')\n","    plt.title(f'Top {top_n} Used Genes in Background for {title}', fontsize=16, fontweight='bold')\n","    plt.xlabel('Background Gene Count', fontsize=14)\n","    plt.ylabel('GO Term', fontsize=14)\n","    plt.xticks(fontsize=12)\n","    plt.yticks(fontsize=12)\n","    sns.despine()  # Removes top and right spines for a cleaner plot\n","\n","    # Save the figure (optional)\n","    plt.savefig(f'../data/plots/{title}_Top_Background_Genes.png', dpi=300, bbox_inches='tight')\n","\n","    plt.show()\n","\n","# Plotting for each DataFrame\n","plot_top_bgd_genes(df1, 'Group1')\n","plot_top_bgd_genes(df2, 'Group2')\n","plot_top_bgd_genes(df3, 'Group3')\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0voQBzQSl3hX"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Molecular functions for the grouping of signal peptides\n","df1 = pd.read_csv('../data/14_GO_analysis/group1_mol_function_GO.tsv', sep='\\t')\n","df2 = pd.read_csv('../data/14_GO_analysis/group2_mol_function_GO.tsv', sep='\\t')\n","df3 = pd.read_csv('../data/14_GO_analysis/group3_mol_function_GO.tsv', sep='\\t')\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"sfn31quLpUqd"},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["# Plotting for each DataFrame\n","plot_top_bgd_genes(df1, 'Group1 - molecular functions')\n","plot_top_bgd_genes(df2, 'Group2 - molecular functions')\n","plot_top_bgd_genes(df3, 'Group3 - molecular functions')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"","evalue":"","output_type":"error","traceback":["\u001b[1;31mRunning cells with '.conda (Python 3.11.7)' requires the ipykernel package.\n","\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n","\u001b[1;31mCommand: 'conda install -p /Users/lucaslevassor/Library/CloudStorage/OneDrive-DanmarksTekniskeUniversitet/projects/Signal_peptide_project/.conda ipykernel --update-deps --force-reinstall'"]}],"source":["synthetic_sps = pd.read_csv('data/15_SecretoGen/Best_signal_peptides_for_A_oryzae_RFP_his_tag_sorted.csv')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"colab":{"authorship_tag":"ABX9TyOYAxoeN5YX498Gdt/ugM0Z","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.12"}},"nbformat":4,"nbformat_minor":0}
