{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data processing\n",
    "\n",
    "### 0 Introduction\n",
    "Before starting data analysis, it is important to questions such as:\n",
    "- What is the outcome to be achieved as a result of data analysis?\n",
    "- Which parameters are necessary to keep, and why, to achieve the desired outcome?\n",
    "- How to achieve a standard structure in different datasets? \n",
    "\n",
    "For this project we have 3 datasets:\n",
    "- **Proteomics:** Signal peptide abundances measured by analyzing the supernatant of fungal samples cultivated under two different media ccompositions (minimal media and minimal media with additional nitrogen)\n",
    "- **RNA sequence:** Gene expression levels obtained from transcriptome anlaysis of fungal samples cultivated under different conditions\n",
    "- **SignalP:** Potential signal peptide seuqences determined by using signalP algorithm to analyze the _A. oryzae_ genome \n",
    "\n",
    "The Proteomics and SingnalP datasets are needed to make comparison between the observed and predicted signal peptides, whereas RNA sequence dataset is needed to normalize the signal peptide abundance obesrved in the proteomics analysis with respect to gene expression levels as this will improve the data for machine learning when we start Automatic Machine Learning (AutoML).\n",
    "\n",
    "However, for now, the datasets have different structure and excess information that are not necessary for data analysis. This will make it difficult to connect the datasets during analysis. Hence, the datasets need to be processed to obtain a common structure which contains only the relevant information needed to carry out analysis and link between the datasets. \n",
    "\n",
    "### Agenda:\n",
    "As a rule of thumb, the datasets should have a generic structure which will satisfy:\n",
    "- Common name for the columns which contain same data content \n",
    "- Representative name that clearly identifies the data stored under the columns\n",
    "- Dataframe strucure that contains only the necessary information needed for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 Data\n",
    "\n",
    "### 1.1 Proteomics data\n",
    "Proteomics data consist of peptide sequence analysis of supernatants obtained from _Aspergillus_ samples that are grown under two different minimum media compositions (minimum media and minimum media with additional nitrogen). Each media composition was tested with 3 samples and the supernatants of each sample were analyzed through mass spectrometry, thus yielding the proteomics dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked</th>\n",
       "      <th>Protein FDR Confidence: Combined</th>\n",
       "      <th>Master</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Exp. q-value: Combined</th>\n",
       "      <th>Sum PEP Score</th>\n",
       "      <th>Coverage [%]</th>\n",
       "      <th># Peptides</th>\n",
       "      <th># PSMs</th>\n",
       "      <th>...</th>\n",
       "      <th>Found in Sample: [S23] F23: Sample, 8</th>\n",
       "      <th>Found in Sample: [S24] F24: Sample, 9</th>\n",
       "      <th>Found in Sample: [S25] F25: Sample, 10</th>\n",
       "      <th>Found in Sample: [S26] F26: Sample, 11</th>\n",
       "      <th>Found in Sample: [S27] F27: Sample, 12</th>\n",
       "      <th>Found in Sample: [S28] F28: Sample, 13</th>\n",
       "      <th>Found in Sample: [S29] F29: Sample, 14</th>\n",
       "      <th>Found in Sample: [S30] F30: Sample, 15</th>\n",
       "      <th># Protein Groups</th>\n",
       "      <th>Modifications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003000935-T-p1</td>\n",
       "      <td>transcript=AO090003000935-T | gene=AO090003000...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>599.687</td>\n",
       "      <td>89</td>\n",
       "      <td>31</td>\n",
       "      <td>7127</td>\n",
       "      <td>...</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090023000944-T-p1</td>\n",
       "      <td>transcript=AO090023000944-T | gene=AO090023000...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>523.784</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>3237</td>\n",
       "      <td>...</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003001591-T-p1</td>\n",
       "      <td>transcript=AO090003001591-T | gene=AO090003001...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>519.742</td>\n",
       "      <td>68</td>\n",
       "      <td>33</td>\n",
       "      <td>2987</td>\n",
       "      <td>...</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>RFP_Fusion</td>\n",
       "      <td>RFP_Fusion</td>\n",
       "      <td>0.000</td>\n",
       "      <td>450.464</td>\n",
       "      <td>64</td>\n",
       "      <td>35</td>\n",
       "      <td>398</td>\n",
       "      <td>...</td>\n",
       "      <td>High</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>High</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>High</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090005001300-T-p1</td>\n",
       "      <td>transcript=AO090005001300-T | gene=AO090005001...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>384.472</td>\n",
       "      <td>90</td>\n",
       "      <td>45</td>\n",
       "      <td>554</td>\n",
       "      <td>...</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090001000075-T-p1</td>\n",
       "      <td>transcript=AO090001000075-T | gene=AO090001000...</td>\n",
       "      <td>0.044</td>\n",
       "      <td>1.795</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090005001355-T-p1</td>\n",
       "      <td>transcript=AO090005001355-T | gene=AO090005001...</td>\n",
       "      <td>0.046</td>\n",
       "      <td>1.790</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090003000247-T-p1</td>\n",
       "      <td>transcript=AO090003000247-T | gene=AO090003000...</td>\n",
       "      <td>0.047</td>\n",
       "      <td>1.779</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>1</td>\n",
       "      <td>Met-loss [N-Term]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090001000680-T-p1</td>\n",
       "      <td>transcript=AO090001000680-T | gene=AO090001000...</td>\n",
       "      <td>0.048</td>\n",
       "      <td>1.774</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>High</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>False</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Master Protein</td>\n",
       "      <td>AO090020000232-T-p1</td>\n",
       "      <td>transcript=AO090020000232-T | gene=AO090020000...</td>\n",
       "      <td>0.048</td>\n",
       "      <td>1.771</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Peak Found</td>\n",
       "      <td>High</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>Not Found</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>866 rows × 87 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Checked Protein FDR Confidence: Combined          Master  \\\n",
       "0      False                             High  Master Protein   \n",
       "1      False                             High  Master Protein   \n",
       "2      False                             High  Master Protein   \n",
       "3      False                             High  Master Protein   \n",
       "4      False                             High  Master Protein   \n",
       "..       ...                              ...             ...   \n",
       "861    False                           Medium  Master Protein   \n",
       "862    False                           Medium  Master Protein   \n",
       "863    False                           Medium  Master Protein   \n",
       "864    False                           Medium  Master Protein   \n",
       "865    False                           Medium  Master Protein   \n",
       "\n",
       "               Accession                                        Description  \\\n",
       "0    AO090003000935-T-p1  transcript=AO090003000935-T | gene=AO090003000...   \n",
       "1    AO090023000944-T-p1  transcript=AO090023000944-T | gene=AO090023000...   \n",
       "2    AO090003001591-T-p1  transcript=AO090003001591-T | gene=AO090003001...   \n",
       "3             RFP_Fusion                                         RFP_Fusion   \n",
       "4    AO090005001300-T-p1  transcript=AO090005001300-T | gene=AO090005001...   \n",
       "..                   ...                                                ...   \n",
       "861  AO090001000075-T-p1  transcript=AO090001000075-T | gene=AO090001000...   \n",
       "862  AO090005001355-T-p1  transcript=AO090005001355-T | gene=AO090005001...   \n",
       "863  AO090003000247-T-p1  transcript=AO090003000247-T | gene=AO090003000...   \n",
       "864  AO090001000680-T-p1  transcript=AO090001000680-T | gene=AO090001000...   \n",
       "865  AO090020000232-T-p1  transcript=AO090020000232-T | gene=AO090020000...   \n",
       "\n",
       "     Exp. q-value: Combined  Sum PEP Score  Coverage [%]  # Peptides  # PSMs  \\\n",
       "0                     0.000        599.687            89          31    7127   \n",
       "1                     0.000        523.784            68          32    3237   \n",
       "2                     0.000        519.742            68          33    2987   \n",
       "3                     0.000        450.464            64          35     398   \n",
       "4                     0.000        384.472            90          45     554   \n",
       "..                      ...            ...           ...         ...     ...   \n",
       "861                   0.044          1.795             4           1       1   \n",
       "862                   0.046          1.790            14           1       1   \n",
       "863                   0.047          1.779             4           1       1   \n",
       "864                   0.048          1.774             1           1       1   \n",
       "865                   0.048          1.771             1           1       2   \n",
       "\n",
       "     ...  Found in Sample: [S23] F23: Sample, 8  \\\n",
       "0    ...                                   High   \n",
       "1    ...                                   High   \n",
       "2    ...                                   High   \n",
       "3    ...                                   High   \n",
       "4    ...                                   High   \n",
       "..   ...                                    ...   \n",
       "861  ...                              Not Found   \n",
       "862  ...                             Peak Found   \n",
       "863  ...                              Not Found   \n",
       "864  ...                              Not Found   \n",
       "865  ...                             Peak Found   \n",
       "\n",
       "     Found in Sample: [S24] F24: Sample, 9  \\\n",
       "0                                     High   \n",
       "1                                     High   \n",
       "2                                     High   \n",
       "3                               Peak Found   \n",
       "4                                     High   \n",
       "..                                     ...   \n",
       "861                             Peak Found   \n",
       "862                             Peak Found   \n",
       "863                             Peak Found   \n",
       "864                              Not Found   \n",
       "865                                   High   \n",
       "\n",
       "     Found in Sample: [S25] F25: Sample, 10  \\\n",
       "0                                      High   \n",
       "1                                      High   \n",
       "2                                 Not Found   \n",
       "3                                      High   \n",
       "4                                      High   \n",
       "..                                      ...   \n",
       "861                              Peak Found   \n",
       "862                              Peak Found   \n",
       "863                              Peak Found   \n",
       "864                              Peak Found   \n",
       "865                               Not Found   \n",
       "\n",
       "     Found in Sample: [S26] F26: Sample, 11  \\\n",
       "0                                      High   \n",
       "1                                Peak Found   \n",
       "2                                 Not Found   \n",
       "3                                Peak Found   \n",
       "4                                      High   \n",
       "..                                      ...   \n",
       "861                              Peak Found   \n",
       "862                               Not Found   \n",
       "863                              Peak Found   \n",
       "864                              Peak Found   \n",
       "865                               Not Found   \n",
       "\n",
       "     Found in Sample: [S27] F27: Sample, 12  \\\n",
       "0                                      High   \n",
       "1                                Peak Found   \n",
       "2                                 Not Found   \n",
       "3                                      High   \n",
       "4                                      High   \n",
       "..                                      ...   \n",
       "861                              Peak Found   \n",
       "862                              Peak Found   \n",
       "863                              Peak Found   \n",
       "864                               Not Found   \n",
       "865                               Not Found   \n",
       "\n",
       "     Found in Sample: [S28] F28: Sample, 13  \\\n",
       "0                                      High   \n",
       "1                                      High   \n",
       "2                                      High   \n",
       "3                                      High   \n",
       "4                                      High   \n",
       "..                                      ...   \n",
       "861                              Peak Found   \n",
       "862                               Not Found   \n",
       "863                               Not Found   \n",
       "864                              Peak Found   \n",
       "865                               Not Found   \n",
       "\n",
       "    Found in Sample: [S29] F29: Sample, 14  \\\n",
       "0                                     High   \n",
       "1                                     High   \n",
       "2                                     High   \n",
       "3                                     High   \n",
       "4                                     High   \n",
       "..                                     ...   \n",
       "861                             Peak Found   \n",
       "862                             Peak Found   \n",
       "863                              Not Found   \n",
       "864                                   High   \n",
       "865                              Not Found   \n",
       "\n",
       "    Found in Sample: [S30] F30: Sample, 15 # Protein Groups      Modifications  \n",
       "0                                     High                1                NaN  \n",
       "1                                     High                1                NaN  \n",
       "2                                     High                1                NaN  \n",
       "3                                     High                1                NaN  \n",
       "4                                     High                1                NaN  \n",
       "..                                     ...              ...                ...  \n",
       "861                             Peak Found                1                NaN  \n",
       "862                              Not Found                1                NaN  \n",
       "863                              Not Found                1  Met-loss [N-Term]  \n",
       "864                             Peak Found                1                NaN  \n",
       "865                              Not Found                1                NaN  \n",
       "\n",
       "[866 rows x 87 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the Excel file into a pandas dataframe \n",
    "df_proteomics = pd.read_excel('/Users/lucaslevassor/projects/Signal_peptide_project/data/03_proteomics_data/20221124_FJ_E1200_MWN_15cm_140min_500ng_#1572_proteins.xlsx')\n",
    "df_proteomics"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 SignalP data\n",
    "\n",
    "SignalP is an algorithm developed in DTU which predicts the presence and location of signal peptide cleavage sites the amino acid sequences through analyzing the annotated protein profile of a specific organism. Hence, SignalP can be used to generate a list of potential signal peptide sequences. I would like to credit and emphasize my gratitude towards my supervisor Lucas Levassor who put the effor to generate the SignalP predictions, which were described in the notebook 00."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>signal_peptide_likelyhood</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AO090005000016-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.999803</td>\n",
       "      <td>MAPSHSFMLFLSVICTHLCSLVVAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>AO090005000029-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.999835</td>\n",
       "      <td>MHLRNIVIALAATAVASPVDLQDRQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>AO090005000042-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.999843</td>\n",
       "      <td>MKASFISRLLSLTAFAISSNLSYGL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>AO090005000053-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.854809</td>\n",
       "      <td>MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVALSQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>AO090005000059-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.999821</td>\n",
       "      <td>MHLQATLAVGLSLLGLTLAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>3168</td>\n",
       "      <td>AO090103000483-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.833106</td>\n",
       "      <td>MKTSFLLAAIGFLYRLPCSAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>3171</td>\n",
       "      <td>AO090103000487-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.999710</td>\n",
       "      <td>MTRYLSFLFLLILFGNSVFTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>3174</td>\n",
       "      <td>AO090103000493-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>MRGIVALSFLSVALGVTAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>3177</td>\n",
       "      <td>AO090701000994-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>MRLLLIAPLFSAVSYGAQAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>3180</td>\n",
       "      <td>AO090701001287-T-p1</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>MLFYRAFGGLLCFLYACVTVSAF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0                 gene  start_pos  end_pos  \\\n",
       "0              0  AO090005000016-T-p1          0       25   \n",
       "1              3  AO090005000029-T-p1          0       25   \n",
       "2              6  AO090005000042-T-p1          0       25   \n",
       "3              9  AO090005000053-T-p1          0       43   \n",
       "4             12  AO090005000059-T-p1          0       20   \n",
       "...          ...                  ...        ...      ...   \n",
       "1056        3168  AO090103000483-T-p1          0       21   \n",
       "1057        3171  AO090103000487-T-p1          0       21   \n",
       "1058        3174  AO090103000493-T-p1          0       19   \n",
       "1059        3177  AO090701000994-T-p1          0       20   \n",
       "1060        3180  AO090701001287-T-p1          0       23   \n",
       "\n",
       "      signal_peptide_likelyhood                                     sequence  \n",
       "0                      0.999803                    MAPSHSFMLFLSVICTHLCSLVVAV  \n",
       "1                      0.999835                    MHLRNIVIALAATAVASPVDLQDRQ  \n",
       "2                      0.999843                    MKASFISRLLSLTAFAISSNLSYGL  \n",
       "3                      0.854809  MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVALSQ  \n",
       "4                      0.999821                         MHLQATLAVGLSLLGLTLAD  \n",
       "...                         ...                                          ...  \n",
       "1056                   0.833106                        MKTSFLLAAIGFLYRLPCSAA  \n",
       "1057                   0.999710                        MTRYLSFLFLLILFGNSVFTA  \n",
       "1058                   0.999791                          MRGIVALSFLSVALGVTAD  \n",
       "1059                   0.999845                         MRLLLIAPLFSAVSYGAQAT  \n",
       "1060                   0.999866                      MLFYRAFGGLLCFLYACVTVSAF  \n",
       "\n",
       "[1061 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the Excel file into a pandas dataframe\n",
    "df_signalP = pd.read_excel('/Users/lucaslevassor/projects/Signal_peptide_project/data/02_all_signal_peptides/sigpep_predict.xlsx')\n",
    "df_signalP"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Data processing\n",
    "\n",
    "What are the parameters to isolate? What to keep and why to keep?\n",
    "\n",
    "### 2.1 Common changes in all datasets\n",
    "\n",
    "Among all the dataset, accession names serve as a common feature that can be used to connect the datasets, thus meaning all the datasets should be processed to have:\n",
    "- Common and representative name for the accession columns such as \"Accession\"\n",
    "- Common data structure which can be recognized such as accession numbers wihtout the suffix: \"-T-p1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the name of the accesion column in the dataframes into \"Accession\" \n",
    "df_signalP = df_signalP.rename(columns={'gene': 'Accession'})\n",
    "\n",
    "# Remove the suffix from the variables located in the \"Accession\" columns\n",
    "df_proteomics['Accession'] = df_proteomics['Accession'].str.replace('-T-p1', '')\n",
    "df_signalP['Accession'] = df_signalP['Accession'].str.replace('-T-p1', '')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Proteomics data processing\n",
    "\n",
    "In this dataset, some of the important processing steps that need to be taken include:\n",
    "\n",
    "- Removal of RFP_Fusion (Red Fluorescent Protein) as it is not part of the native protein pool of A. oryzae RIB40\n",
    "- Isolation of important parameters for analysis such as the columns:\n",
    "    - **Accession:** Unique identifier that is assigned to the signal peptide sequence \n",
    "    - **Abundace:** Raw signal peptide counts that are observed and recorded directly from mass spectrometry analysis \n",
    "    - **Abundances (Scaled):** Scaled version of raw singal peptide counts signal peptide counts that are observed from mass spectrometry analysis \n",
    "\n",
    "There are multiple columns with the name \"Abundance\" hence, distinctive patterns needs to be determined to isolate the specific columns. The data for minimum media (MM) and minimum media with additional nitrogen (MM_N) compositions are stored under the column names which show a specific naming pattern such as:\n",
    "\n",
    "- **Raw signal peptide abundance**\n",
    "    - Abundance: F16, Abundance: F17, Abundance: F18 - Minimum media (MM)\n",
    "    - Abundance: F28, Abundance: F29, Abundance: F30 - Minimum media with additional nitrogen (MM_N)\n",
    "- **Scaled signal peptide abundance**\n",
    "    - Abundances (Scaled): F16, Abundances (Scaled): F17, Abundances (Scaled): F18 - Minimum media (MM)\n",
    "    - Abundances (Scaled): F28, Abundances (Scaled): F29, Abundances (Scaled): F30 - Minimum media with additional nitrogen (MM_N)\n",
    "\n",
    "Hence, the \"Scaled\" and \"Raw\" abundace samples can be distinguished with the column pattern names:\n",
    "- **Pattern 1:** \"Abundance: and Scaled\"\n",
    "- **Pattern 2:** \"F16, F17, F18, F28, F29, F30\"\n",
    "\n",
    "Lets remove \"RFP_Fusion\" and create a boolean mask to select dataframe variables according to specific characters they contain in the dataframe such as using \"Scaled\" to select scaled abundances and \"Abundance:\" to select raw abundances "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Abundances (Scaled): F16: Sample, 1</th>\n",
       "      <th>Abundances (Scaled): F17: Sample, 2</th>\n",
       "      <th>Abundances (Scaled): F18: Sample, 3</th>\n",
       "      <th>Abundances (Scaled): F28: Sample, 13</th>\n",
       "      <th>Abundances (Scaled): F29: Sample, 14</th>\n",
       "      <th>Abundances (Scaled): F30: Sample, 15</th>\n",
       "      <th>Abundance: F16: Sample, 1</th>\n",
       "      <th>Abundance: F17: Sample, 2</th>\n",
       "      <th>Abundance: F18: Sample, 3</th>\n",
       "      <th>Abundance: F28: Sample, 13</th>\n",
       "      <th>Abundance: F29: Sample, 14</th>\n",
       "      <th>Abundance: F30: Sample, 15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090003000935</td>\n",
       "      <td>5.4</td>\n",
       "      <td>4.3</td>\n",
       "      <td>7.4</td>\n",
       "      <td>81.0</td>\n",
       "      <td>35.7</td>\n",
       "      <td>77.8</td>\n",
       "      <td>9.597858e+07</td>\n",
       "      <td>9.416572e+07</td>\n",
       "      <td>1.297344e+08</td>\n",
       "      <td>1.888527e+09</td>\n",
       "      <td>1.011602e+09</td>\n",
       "      <td>1.584488e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090023000944</td>\n",
       "      <td>242.7</td>\n",
       "      <td>250.1</td>\n",
       "      <td>258.5</td>\n",
       "      <td>42.9</td>\n",
       "      <td>72.6</td>\n",
       "      <td>45.7</td>\n",
       "      <td>2.378266e+08</td>\n",
       "      <td>3.016177e+08</td>\n",
       "      <td>2.504219e+08</td>\n",
       "      <td>5.514791e+07</td>\n",
       "      <td>1.134059e+08</td>\n",
       "      <td>5.128523e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090003001591</td>\n",
       "      <td>231.8</td>\n",
       "      <td>214.1</td>\n",
       "      <td>244.3</td>\n",
       "      <td>62.8</td>\n",
       "      <td>84.6</td>\n",
       "      <td>60.1</td>\n",
       "      <td>9.888853e+09</td>\n",
       "      <td>1.123887e+10</td>\n",
       "      <td>1.030518e+10</td>\n",
       "      <td>3.511919e+09</td>\n",
       "      <td>5.755309e+09</td>\n",
       "      <td>2.932830e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090005001300</td>\n",
       "      <td>96.3</td>\n",
       "      <td>38.4</td>\n",
       "      <td>17.0</td>\n",
       "      <td>89.7</td>\n",
       "      <td>111.3</td>\n",
       "      <td>82.3</td>\n",
       "      <td>7.204276e+07</td>\n",
       "      <td>3.532077e+07</td>\n",
       "      <td>1.257048e+07</td>\n",
       "      <td>8.793645e+07</td>\n",
       "      <td>1.328252e+08</td>\n",
       "      <td>7.046364e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090010000746</td>\n",
       "      <td>98.6</td>\n",
       "      <td>105.3</td>\n",
       "      <td>113.0</td>\n",
       "      <td>260.9</td>\n",
       "      <td>284.5</td>\n",
       "      <td>247.8</td>\n",
       "      <td>2.367148e+08</td>\n",
       "      <td>3.107989e+08</td>\n",
       "      <td>2.681958e+08</td>\n",
       "      <td>8.206836e+08</td>\n",
       "      <td>1.088923e+09</td>\n",
       "      <td>6.807444e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>AO090001000075</td>\n",
       "      <td>201.7</td>\n",
       "      <td>18.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>239.4</td>\n",
       "      <td>431.0</td>\n",
       "      <td>261.4</td>\n",
       "      <td>1.819305e+05</td>\n",
       "      <td>2.070978e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.830392e+05</td>\n",
       "      <td>6.199164e+05</td>\n",
       "      <td>2.698529e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>AO090005001355</td>\n",
       "      <td>64.6</td>\n",
       "      <td>146.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>184.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.104682e+04</td>\n",
       "      <td>5.864526e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.575187e+04</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>AO090003000247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>AO090001000680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>250.9</td>\n",
       "      <td>673.0</td>\n",
       "      <td>244.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.761538e+04</td>\n",
       "      <td>1.879930e+05</td>\n",
       "      <td>4.894831e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>AO090020000232</td>\n",
       "      <td>265.1</td>\n",
       "      <td>263.7</td>\n",
       "      <td>267.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.173873e+05</td>\n",
       "      <td>2.661302e+05</td>\n",
       "      <td>2.164992e+05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>865 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Accession  Abundances (Scaled): F16: Sample, 1  \\\n",
       "0    AO090003000935                                  5.4   \n",
       "1    AO090023000944                                242.7   \n",
       "2    AO090003001591                                231.8   \n",
       "4    AO090005001300                                 96.3   \n",
       "5    AO090010000746                                 98.6   \n",
       "..              ...                                  ...   \n",
       "861  AO090001000075                                201.7   \n",
       "862  AO090005001355                                 64.6   \n",
       "863  AO090003000247                                  NaN   \n",
       "864  AO090001000680                                  NaN   \n",
       "865  AO090020000232                                265.1   \n",
       "\n",
       "     Abundances (Scaled): F17: Sample, 2  Abundances (Scaled): F18: Sample, 3  \\\n",
       "0                                    4.3                                  7.4   \n",
       "1                                  250.1                                258.5   \n",
       "2                                  214.1                                244.3   \n",
       "4                                   38.4                                 17.0   \n",
       "5                                  105.3                                113.0   \n",
       "..                                   ...                                  ...   \n",
       "861                                 18.7                                  NaN   \n",
       "862                                146.2                                  NaN   \n",
       "863                                  NaN                                  NaN   \n",
       "864                                  NaN                                  NaN   \n",
       "865                                263.7                                267.0   \n",
       "\n",
       "     Abundances (Scaled): F28: Sample, 13  \\\n",
       "0                                    81.0   \n",
       "1                                    42.9   \n",
       "2                                    62.8   \n",
       "4                                    89.7   \n",
       "5                                   260.9   \n",
       "..                                    ...   \n",
       "861                                 239.4   \n",
       "862                                   NaN   \n",
       "863                                   NaN   \n",
       "864                                 250.9   \n",
       "865                                   NaN   \n",
       "\n",
       "     Abundances (Scaled): F29: Sample, 14  \\\n",
       "0                                    35.7   \n",
       "1                                    72.6   \n",
       "2                                    84.6   \n",
       "4                                   111.3   \n",
       "5                                   284.5   \n",
       "..                                    ...   \n",
       "861                                 431.0   \n",
       "862                                 184.2   \n",
       "863                                   NaN   \n",
       "864                                 673.0   \n",
       "865                                   NaN   \n",
       "\n",
       "     Abundances (Scaled): F30: Sample, 15  Abundance: F16: Sample, 1  \\\n",
       "0                                    77.8               9.597858e+07   \n",
       "1                                    45.7               2.378266e+08   \n",
       "2                                    60.1               9.888853e+09   \n",
       "4                                    82.3               7.204276e+07   \n",
       "5                                   247.8               2.367148e+08   \n",
       "..                                    ...                        ...   \n",
       "861                                 261.4               1.819305e+05   \n",
       "862                                   NaN               2.104682e+04   \n",
       "863                                   NaN                        NaN   \n",
       "864                                 244.1                        NaN   \n",
       "865                                   NaN               2.173873e+05   \n",
       "\n",
       "     Abundance: F17: Sample, 2  Abundance: F18: Sample, 3  \\\n",
       "0                 9.416572e+07               1.297344e+08   \n",
       "1                 3.016177e+08               2.504219e+08   \n",
       "2                 1.123887e+10               1.030518e+10   \n",
       "4                 3.532077e+07               1.257048e+07   \n",
       "5                 3.107989e+08               2.681958e+08   \n",
       "..                         ...                        ...   \n",
       "861               2.070978e+04                        NaN   \n",
       "862               5.864526e+04                        NaN   \n",
       "863                        NaN                        NaN   \n",
       "864                        NaN                        NaN   \n",
       "865               2.661302e+05               2.164992e+05   \n",
       "\n",
       "     Abundance: F28: Sample, 13  Abundance: F29: Sample, 14  \\\n",
       "0                  1.888527e+09                1.011602e+09   \n",
       "1                  5.514791e+07                1.134059e+08   \n",
       "2                  3.511919e+09                5.755309e+09   \n",
       "4                  8.793645e+07                1.328252e+08   \n",
       "5                  8.206836e+08                1.088923e+09   \n",
       "..                          ...                         ...   \n",
       "861                2.830392e+05                6.199164e+05   \n",
       "862                         NaN                9.575187e+04   \n",
       "863                         NaN                         NaN   \n",
       "864                5.761538e+04                1.879930e+05   \n",
       "865                         NaN                         NaN   \n",
       "\n",
       "     Abundance: F30: Sample, 15  \n",
       "0                  1.584488e+09  \n",
       "1                  5.128523e+07  \n",
       "2                  2.932830e+09  \n",
       "4                  7.046364e+07  \n",
       "5                  6.807444e+08  \n",
       "..                          ...  \n",
       "861                2.698529e+05  \n",
       "862                         NaN  \n",
       "863                         NaN  \n",
       "864                4.894831e+04  \n",
       "865                         NaN  \n",
       "\n",
       "[865 rows x 13 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove RFP_Fusion from the dataframe as it is not part of the native protein pool\n",
    "df_proteomics = df_proteomics[df_proteomics.Accession != 'RFP_Fusion']\n",
    "\n",
    "# Create a boolean mask that is True for columns that contain the pattaern1: \"Scaled, Abundance:\" and pattern2: \"F16, F17, F18, F28, F29; F30\"\n",
    "pattern1 = 'Abundance:|Scaled'\n",
    "pattern2 = 'F16|F17|F18|F28|F29|F30'\n",
    "df_raw_scaled_abundances = df_proteomics.loc[:, df_proteomics.columns.str.contains(pattern1) & df_proteomics.columns.str.contains(pattern2, regex=True)]\n",
    "\n",
    "# Recall the dataframe to combine accession with ioslated list of columns for the peptide abundances \n",
    "df_proteomics_processed = df_proteomics[['Accession']+ list(df_raw_scaled_abundances.columns)]\n",
    "df_proteomics_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and save a summary dataframe\n",
    "df_proteomics_processed.to_csv('/Users/lucaslevassor/projects/Signal_peptide_project/data/03_proteomics_data/proteomics_processed.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 SignalP data processing\n",
    "\n",
    "In this dataset, some of the important processing steps that need to be taken include:\n",
    "- Change the column names \"end_pos\" to a representative name\n",
    "- Isolation of important parameters for analysis such as the columns:\n",
    "    - **Accession:** Unique identifier that is assigned to the signal peptide sequence \n",
    "    - **end_pos:** Length of the predicted signal peptides\n",
    "    - **sequence:** Amino acid sequence of the predicfted signal peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090005000016</td>\n",
       "      <td>25</td>\n",
       "      <td>MAPSHSFMLFLSVICTHLCSLVVAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090005000029</td>\n",
       "      <td>25</td>\n",
       "      <td>MHLRNIVIALAATAVASPVDLQDRQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005000042</td>\n",
       "      <td>25</td>\n",
       "      <td>MKASFISRLLSLTAFAISSNLSYGL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090005000053</td>\n",
       "      <td>43</td>\n",
       "      <td>MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVALSQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090005000059</td>\n",
       "      <td>20</td>\n",
       "      <td>MHLQATLAVGLSLLGLTLAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>AO090103000483</td>\n",
       "      <td>21</td>\n",
       "      <td>MKTSFLLAAIGFLYRLPCSAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>AO090103000487</td>\n",
       "      <td>21</td>\n",
       "      <td>MTRYLSFLFLLILFGNSVFTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>AO090103000493</td>\n",
       "      <td>19</td>\n",
       "      <td>MRGIVALSFLSVALGVTAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>AO090701000994</td>\n",
       "      <td>20</td>\n",
       "      <td>MRLLLIAPLFSAVSYGAQAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>AO090701001287</td>\n",
       "      <td>23</td>\n",
       "      <td>MLFYRAFGGLLCFLYACVTVSAF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Accession  length                                     sequence\n",
       "0     AO090005000016      25                    MAPSHSFMLFLSVICTHLCSLVVAV\n",
       "1     AO090005000029      25                    MHLRNIVIALAATAVASPVDLQDRQ\n",
       "2     AO090005000042      25                    MKASFISRLLSLTAFAISSNLSYGL\n",
       "3     AO090005000053      43  MGLFLTALGALSSVNVLYSRGRMPLKHLATLLCALSPTVALSQ\n",
       "4     AO090005000059      20                         MHLQATLAVGLSLLGLTLAD\n",
       "...              ...     ...                                          ...\n",
       "1056  AO090103000483      21                        MKTSFLLAAIGFLYRLPCSAA\n",
       "1057  AO090103000487      21                        MTRYLSFLFLLILFGNSVFTA\n",
       "1058  AO090103000493      19                          MRGIVALSFLSVALGVTAD\n",
       "1059  AO090701000994      20                         MRLLLIAPLFSAVSYGAQAT\n",
       "1060  AO090701001287      23                      MLFYRAFGGLLCFLYACVTVSAF\n",
       "\n",
       "[1061 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename the column \"Unnamed: 5\" into a representative name\n",
    "df_signalP = df_signalP.rename(columns={'end_pos': 'length'})\n",
    "\n",
    "# Isolate the important parameters\n",
    "df_signalP_processed = df_signalP[['Accession', 'length', 'sequence']]\n",
    "df_signalP_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv_spep",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
